File: MergeKSortedList/mergeKSortedList.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode[]} lists
 * @return {ListNode}
 */

class ListNode {
  constructor(val, next) {
    this.val = (val === undefined ? 0 : val)
    this.next = (next === undefined ? null : next)
  }
}

class MinHeap {
  constructor() {
    this.heap = [];
  }

  show() {
    console.log(this.heap);
    return;
  }

  insert(node) {
    this.heap.push(node);
    this.bubbleUp(this.heap.length - 1);
  }

  extractMin() {

    if (this.heap.length === 0) { return null; }
    if (this.heap.length === 1) { return this.heap.pop(); }
    let min = this.heap[0];
    this.heap.shift();
    this.bubbleDown(0);
    return min;
  }

  bubbleUp(index) {
    while (index > 0) {
      if (this.heap[index].val < this.heap[index - 1].val) {
        [this.heap[index], this.heap[index - 1]] = [this.heap[index - 1], this.heap[index]];
        index -= 1;
      } else {
        break;
      }
    }
  }

  bubbleDown(index) {
    let isFinished = this.heap.length === 1 ? true : false;
    while (!isFinished) {
      if (this.heap[index].val > this.heap[index + 1].val && index < this.heap.length - 1) {
        [this.heap[index], this.heap[index + 1]] = [this.heap[index + 1], this.heap[index]];
        index += 1;
      } else {
        isFinished = true;
        break;
      }
    }
    return;
  }

  isEmpty() {
    return this.heap.length === 0;
  }
}

var mergeKLists = function(lists) {
  const minHeap = new MinHeap();

  lists.forEach(element => {
    if (element != null) {
      minHeap.insert(element);
    }
  });

  let min;
  let mergedList = new ListNode();
  let current = mergedList;
  while (!minHeap.isEmpty()) {
    min = minHeap.extractMin();
    current.next = min;
    current = current.next;

    if (min.next != null) {
      minHeap.insert(min.next);
    }
  }
  return mergedList.next;
};

const list1 = new ListNode(1, new ListNode(4, new ListNode(5)));
const list2 = new ListNode(1, new ListNode(3, new ListNode(4)));
const list3 = new ListNode(2, new ListNode(6));

const data = [list1, list2, list3]

let answer;
answer = mergeKLists(data);

console.log(answer.val);
while (answer.next != null) {
  answer = answer.next;
  console.log(answer.val);
}

File: MergeKSortedList/mergeKSortedList.py
import math


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class MinHeap():

    def __init__(self):
        self.heap = []

    def show(self):
        return self.heap

    def insert(self, val):
        self.heap.append(val)
        if len(self.heap) == 1:
            return
        self.bubbleUp()
        return
    
    def bubbleUp(self):
        indexAdded = len(self.heap) - 1
        while indexAdded > 0:
            parentIndex = math.floor((indexAdded - 1) / 2)
            if (self.heap[indexAdded] < self.heap[parentIndex]):
                self.heap[indexAdded], self.heap[parentIndex] = self.heap[parentIndex], self.heap[indexAdded]
                indexAdded = parentIndex
            else:
                break
        return
    
    def getMinValue(self):
        if len(self.heap) == 0:
            return
        if len(self.heap) == 1:
            return self.heap.pop()
        
        minValue = self.heap[0]
        self.heap[0] = self.heap.pop()
        self.bubbleDown(0)
        return minValue
    
    def bubbleDown(self, index):
        heapLength = len(self.heap)
        while self.hasLeftNode(index):
            leftNode = (index * 2) + 1
            rightNode = (index * 2) + 2
            minValue = leftNode
            if rightNode < heapLength:
                minValue = rightNode if self.heap[rightNode] < self.heap[minValue] else minValue
            if self.heap[minValue] < self.heap[index]:
                self.heap[minValue], self.heap[index] = self.heap[index], self.heap[minValue]
                index = minValue
            else:
                break
        return
    
    def hasLeftNode(self, index):
        return (index * 2) + 1 < len(self.heap)

    def isEmpty(self):
        return True if len(self.heap) <= 0 else False



class Solution:
    def mergeKLists(self, lists):
        minHeap = MinHeap()
        for items in lists:
            flag = items
            while flag:
                minHeap.insert(flag.val)
                flag = flag.next
        orderArr = ListNode()
        current = orderArr
        while minHeap.isEmpty() != True:
            current.next = ListNode(minHeap.getMinValue())
            current = current.next
        return orderArr.next


list1 = ListNode(1,ListNode(4, ListNode(5)))
list2 = ListNode(1,ListNode(3, ListNode(4)))
list3 = ListNode(2,ListNode(6))

# list1 = ListNode(1,ListNode(2, ListNode(2)))
# list2 = ListNode(1,ListNode(1, ListNode(2)))
lists = [list1, list2, list3]
sol = Solution()

def printSolution(list):
    while list:
        print(list.val, end="->")
        list = list.next
    print("None")

printSolution(sol.mergeKLists(lists))
